### PR TITLE
Fix/idempotent umount for slink not exist and PV not exist

### PR DESCRIPTION
**Pull request by shay-berman**
This PR is to fix some idempotent issues:
1. **fix idempotent 1: if PV not exist** in ubiqutiy DB then finish unmount with success, assuming the volume is no longer exist or mapped.
(Coding aspects: Flex identify the PV not exist in DB by using generic error comes from https://github.com/IBM/ubiquity/pull/216)
2. **fix idempotent 2: if slink (/k8s-pod-dir/PV) not exist** during unmount (delete Pod) then skip
 umount operation(since its was already unmounted) and go to detach stage of the unmount (no need to run umount nor delete the slink). 
 (Coding aspects : Added checkSlinkBeforeUmount() to identify the idempotent state and other abnormal slink states, like if the slink is a different type or some error to identify it.)
3. fix idempotent 3 if slink (/k8s-pod-dir/PV) **point to wrong mountpoint**  or the moutpoint given is not slink then raise error.

**In addition some refactoring** in the flex controller.Unmount():
1. **Identify backend type** by GetVolume instead of hardcoded check by the mountpath.  (coding : getRealMountpointForPvByBackend func)
2. **Agnostic flow for SCBE and Ssc backends** (dropped doUnmountSsc and doUnmountScbe, instead do the same flow for both)
3. **Reuse code** failureFlexVolumeResponse() instead to build it each time.
This PR related to internal ticket UB-1012
4. Adding full unit tests for this PR (there was none on Unmount func).

---
This change is [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)

### CODE REVIEW

**Comment by olgashtivelman** [comment]
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[glide.yaml, line 12 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcc8_D8jWE9myYS1yM:-LEcc8_D8jWE9myYS1yN:bbtmrki) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/glide.yaml#L12)):*
> ```YAML
>   - lib
> - package: github.com/IBM/ubiquity
>   version: fix/idempotent_umount_improvments
> ```

should this not always be "dev" for the dev branch? 

---

*[controller/controller.go, line 184 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEccbs5478r4lWSNUjB:-LEccbs5478r4lWSNUjC:b-plxqgj) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/controller.go#L184)):*
> ```Go
> 	} else {
> 		response = k8sresources.FlexVolumeResponse{
> 			Status:   "Success",
> ```

did you not want to change all the succeessVolumeResponses to the new c.successFlexVolumeResponse type? (I see that you did it above but from here down i don't see you using it)
(if so there is another one that you can change to the new successResonse a few lines bellow)

---

*[controller/controller.go, line 313 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcef1Z-QY9Mzg_gBLm:-LEcePpL3TBNttgUgejI:byjc4f7) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/controller.go#L313)):*
> ```Go
> 		if c.exec.IsNotExist(err) {
> 			// The k8s PV directory not exist (its a rare case and indicate on idempotent flow)
> 			c.logger.Info("PV directory(k8s-mountpoint) is not exist.", logs.Args{{"k8s-mountpoint", k8sPVDirectoryPath}, {"should-point-to-mountpoint", realMountedPath}}) // TODO change it to warning
> ```

should be "does not exist" :) 

---

*[controller/controller.go, line 431 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcloIy-xxPysbEeiKH:-LEcloIy-xxPysbEeiKI:b-kvfbgn) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/controller.go#L431)):*
> ```Go
> 	}
> 
> 	if err := c.doUnmount(k8sPVDirectoryPath, volume.Backend, volumeConfig, mounter); err != nil {
> ```

so before as far as i see there was an option according to the backend to either call doUnmountSSc or doUnmountScbe but now you are only calliong the new doUnount fucntion and I see that you deleted the doUnmountScbe, but the doUnmountSSc is still here in the code, is it still needed if you are not calling it here anymore? 

---

*[controller/errors.go, line 103 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcn1p16RE7Vrb7zHUH:-LEcn1p16RE7Vrb7zHUI:b7yulhs) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/errors.go#L103)):*
> ```Go
> 
> func (e *PvBackendNotSupportedError) Error() string {
> 	// The error string contains also the fileInfo for debug purpose, in order to identify for example what is the actual FileInfo.Mode().
> ```

copy\paster? I guess you wanted to say that the error contains the back-end ?

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by olgashtivelman** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by shay-berman** [comment]
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[glide.yaml, line 12 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcc8_D8jWE9myYS1yM:-LEdYZT58Hy2BIQTM6jj:b-h56n7m) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/glide.yaml#L12)):*
<details><summary><i>Previously, olgashtivelman wrote…</i></summary><blockquote>

should this not always be "dev" for the dev branch? 
</blockquote></details>

yes
but for testing it with the new generic error that comes from ubiqutiy repo I had to specify this branch. This is how i run the sanity on it.   Before merge to dev I will change this version to dev.

---

*[controller/controller.go, line 184 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEccbs5478r4lWSNUjB:-LEdYqeT5u4M5tuSxtJH:b-7yo68w) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/controller.go#L184)):*
<details><summary><i>Previously, olgashtivelman wrote…</i></summary><blockquote>

did you not want to change all the succeessVolumeResponses to the new c.successFlexVolumeResponse type? (I see that you did it above but from here down i don't see you using it)
(if so there is another one that you can change to the new successResonse a few lines bellow)
</blockquote></details>

Yes, since it was not straight foreword to change all the Success response I didn't change it. Only the failure response.
On a different PR we should refactor also the Success response. I am trying to keep this PR small as possible.

---

*[controller/controller.go, line 313 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcef1Z-QY9Mzg_gBLm:-LEdZPLm2ESURZNjlK8Q:b-863nt) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/controller.go#L313)):*
<details><summary><i>Previously, olgashtivelman wrote…</i></summary><blockquote>

should be "does not exist" :) 
</blockquote></details>

you right - fixed.

---

*[controller/controller.go, line 431 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcloIy-xxPysbEeiKH:-LEdZszF5EPVlTh56Wmf:b-ub6kiy) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/controller.go#L431)):*
<details><summary><i>Previously, olgashtivelman wrote…</i></summary><blockquote>

so before as far as i see there was an option according to the backend to either call doUnmountSSc or doUnmountScbe but now you are only calliong the new doUnount fucntion and I see that you deleted the doUnmountScbe, but the doUnmountSSc is still here in the code, is it still needed if you are not calling it here anymore? 
</blockquote></details>

you right, but I didn't delete it for a reason:
I didn't delete the doUnmountSSc because I wanted Yadav(put him as reviewer to the PR)  to review this change, and make sure Its generic enough also for Spectrum Scale.   If so then we can delete this doUnmountSSc function.
But i want Yadav to confirm that.

---

*[controller/errors.go, line 103 at r1](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEcn1p16RE7Vrb7zHUH:-LEd_hQu54XOLXS4W8Yp:b-9wqcqr) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/ae0835cdfb8c660e181c685d79c76bc568cb1ccd/controller/errors.go#L103)):*
<details><summary><i>Previously, olgashtivelman wrote…</i></summary><blockquote>

copy\paster? I guess you wanted to say that the error contains the back-end ?
</blockquote></details>

good catch, thanks!  
Fixed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by shay-berman** [comment]
@olgashtivelman thanks for code review comments, please review again per my fixes.

**Comment by olgashtivelman** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by olgashtivelman** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-:-LEdxPaBEI0-Jq0WHsnk:bnfp4nl)*

File: controller/controller.go [hunk 1]
```diff
@@ -294,11 +274,122 @@ func (c *Controller) Mount(mountRequest k8sresources.FlexVolumeMountRequest) k8s
...
+	}
+
+}
+
+func (c *Controller) getRealMountpointForPvByBackend(volumeBackend string, volumeConfig map[string]interface{}) (string, error) {
+	// TODO we should create agnostic function based on backend that returns the real mountpoint
+	defer c.logger.Trace(logs.DEBUG)()
+	if volumeBackend == resources.SCBE {
+		return fmt.Sprintf(resources.PathToMountUbiquityBlockDevices, volumeConfig["Wwn"].(string)), nil
+	} else if volumeBackend == resources.SpectrumScale {
```
Comment (gaurang1988) [inline controller/controller.go]: I believe this would always return empty string for Spectrum Scale backend and unmount would fail. This has to be taken care.

File: controller/controller.go [hunk 2]
```diff
@@ -294,11 +274,122 @@ func (c *Controller) Mount(mountRequest k8sresources.FlexVolumeMountRequest) k8s
...
+		return "", &BackendNotImplementedGetRealMountpointError{Backend: volumeBackend}
```
Comment (gaurang1988) [inline controller/controller.go]: I believe this would always return empty string as realMountPoint and would always fail for Spectrum Scale. Need to fix this.

File: controller/controller.go [hunk 3]
```diff
@@ -294,11 +274,122 @@ func (c *Controller) Mount(mountRequest k8sresources.FlexVolumeMountRequest) k8s
...
+	} else {
+		return "", &PvBackendNotSupportedError{Backend: volumeBackend}
```
Comment (gaurang1988) [inline controller/controller.go]: Need to consider a case for Spectrum Scale NFS backend as well.

**Comment by shay-berman** [comment]
Review status: all files reviewed, 2 unresolved discussions (waiting on @shay-berman and @yadaven)

---

*[controller/controller.go, line 346 at r2](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEPCviiK1Uxps--HH_5-r2-346:-LFv-5413vszWldHkWRj:b-ygpy0z) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/f8d38f7d6fd8f679ae4497e91c5eeea213b534af/controller/controller.go#L346)):*
<details><summary><i>Previously, gaurang1988 (Gaurang Tapase) wrote…</i></summary><blockquote>

I believe this would always return empty string as realMountPoint and would always fail for Spectrum Scale. Need to fix this.
</blockquote></details>

yes I know, its should be implemented by SSc. For now it raise NotImplemented error.   This is thing you will have to fill up according to scale need (in doUnmountSsc it save the mountpoint inside the ubiqutiy DB but it doesn't make sense for saving the mountpoint in the DB, because if u will use the same PV on different cluster the mountpoint may be different).  
Anyway we should consider to refactor this deeper so we will not have switch case here per backend. We should do it more elegant way like additional interface in the mounter or generic mounter in the flex side it self.

But for now I leave it like that.  Please remember to fill it up with Scale requirement in the PR you will propose later on inside the github.
Do you agree?

---

*[controller/controller.go, line 348 at r2](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEPCviiK1Uxps--HH_5-r2-348:-LFv-jJo8bi7psn8YuCY:b-5d44r6) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/f8d38f7d6fd8f679ae4497e91c5eeea213b534af/controller/controller.go#L348)):*
<details><summary><i>Previously, gaurang1988 (Gaurang Tapase) wrote…</i></summary><blockquote>

Need to consider a case for Spectrum Scale NFS backend as well.
</blockquote></details>

sure 
same thing as mentioned above.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by shay-berman** [comment]
Review status: all files reviewed, 2 unresolved discussions (waiting on @shay-berman, @gaurang1988, and @yadaven)

---

*[controller/controller.go, line 346 at r2](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-LEPCviiK1Uxps--HH_5-r2-346:-LFv6nifC3fLK_xlmRoM:b-58h2mz) ([raw file](https://github.com/ibm/ubiquity-k8s/blob/f8d38f7d6fd8f679ae4497e91c5eeea213b534af/controller/controller.go#L346)):*
<details><summary><i>Previously, shay-berman wrote…</i></summary><blockquote>

yes I know, its should be implemented by SSc. For now it raise NotImplemented error.   This is thing you will have to fill up according to scale need (in doUnmountSsc it save the mountpoint inside the ubiqutiy DB but it doesn't make sense for saving the mountpoint in the DB, because if u will use the same PV on different cluster the mountpoint may be different).  
Anyway we should consider to refactor this deeper so we will not have switch case here per backend. We should do it more elegant way like additional interface in the mounter or generic mounter in the flex side it self.

But for now I leave it like that.  Please remember to fill it up with Scale requirement in the PR you will propose later on inside the github.
Do you agree?
</blockquote></details>

In addition, be aware that the doUnmountSsc is doing detach but it should not do nothing because its unmount operation and not detach operation.  Anyway the code in DEV is not compatible with scale yet.
You need to create PR that fix the framework for SSc, this is the first thing you need to do.

---


*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by olgashtivelman** [comment]
:lgtm:

---

Review status: 1 of 3 files reviewed, 2 unresolved discussions (waiting on @olgashtivelman, @gaurang1988, and @yadaven)

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193#-:-LFwJ8GBAyFEQQqJZlAh:bnfp4nl)*

**Comment by olgashtivelman** [comment]
Reviewed 1 of 2 files at r3.
Review status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @gaurang1988, @olgashtivelman, and @yadaven)

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*

**Comment by olgashtivelman** [comment]
Reviewed 1 of 4 files at r1.
Review status: all files reviewed, 2 unresolved discussions (waiting on @gaurang1988 and @yadaven)

---



*Comments from [Reviewable](https://reviewable.io/reviews/ibm/ubiquity-k8s/193)*
